Setup Activity components

NahomAlemu opened the pull request:
- Added ActivityController with CRUD operations and exception handling
- Implemented ActivityService to encapsulate the business logic
- Created an Activity entity to map the activity table
- Added ActivityRepository for database operations
- Added custom exceptions for activity-related operations
- Added 'user' argument into saveGoal and createGoal methods
- Added 'getGoalById' into GoalService
Issue:

Closes #27

NahomAlemu committed:
Setup Activity components

- Added ActivityController with CRUD operations and exception handling
- Implemented ActivityService to encapsulate the business logic
- Created Activity entity to map the activity table
- Added ActivityRepository for database operations
- Added custom exceptions for activity-related operations

Issue:

Closes #27

NahomAlemu committed:
Setup Activity components

- Added ActivityController with CRUD operations and exception handling
- Implemented ActivityService to encapsulate the business logic
- Created Activity entity to map the activity table
- Added ActivityRepository for database operations
- Added custom exceptions for activity-related operations

Issue:

Closes #27

CourtneyHoppus reviewed (commented):
When updating an activity, you are checking that the request body activity has an id and that it matches the id for the activity in the database. We aren’t doing this when we update the User or the Goal, I was wondering what your thoughts were on how we are capturing that ID and passing that in. I can do it manually in a Postman test, but that’s only for that. Plus the way it’s written, you can only update the description, we also need to be able to update the title, isComplete, and possibly type.

I was able to create two activities for two different users, but I didn’t change the goal id number and it still worked. So I created an activity for users/4/activities/3 and users/2/activities/3 and they were both created for the correct user, but the goal id 3, which is a user 4 goal, didn’t seem to come into play. I can get both users’ activities, they both have one. I was also able to delete the User 2 activity, again this is the one that the Goal was not for that User.

NahomAlemu committed:
Merge remote-tracking branch 'origin/feature/setup-initial-activity' into feature/setup-initial-activity

NahomAlemu commented:
> When updating an activity, you are checking that the request body activity has an id and that it matches the id for the activity in the database. We aren’t doing this when we update the User or the Goal, I was wondering what your thoughts were on how we are capturing that ID and passing that in. I can do it manually in a Postman test, but that’s only for that. Plus the way it’s written, you can only update the description, we also need to be able to update the title, isComplete, and possibly type.
> 
> I was able to create two activities for two different users, but I didn’t change the goal id number and it still worked. So I created an activity for users/4/activities/3 and users/2/activities/3 and they were both created for the correct user, but the goal id 3, which is a user 4 goal, didn’t seem to come into play. I can get both users’ activities, they both have one. I was also able to delete the User 2 activity, again this is the one that the Goal was not for that User.

@CourtneyHoppus I appreciate your keen observations.

Regarding your first point about updating activities:

You're correct. Adding the ID in the request body was redundant and inconsistent with how we update the User and Goal. I've removed it.
I've also updated the method to allow for updating other fields like title, isComplete, and type along with the description.

On to your second point about creating activities for different users:

The issue seems to lie in the Goal component. We need to add a similar method to findActivityByIdAndUser, but for goals. Something like findGoalByIdAndUser. This method should fetch the goal by its ID and check that it belongs to the user, throwing an exception if either of these checks fails.
This will address the issue where an activity could be created under a goal that doesn't belong to the user. 

Would you be able to implement the findGoalByIdAndUser method in your Goal component?

NahomAlemu commented:
This branch has been superseded by a new branch to address the issues more effectively. Closing this pull request, the branch will be kept temporarily for reference.
